### PR TITLE
improved error and response handling

### PR DESCRIPTION
**Pull request by richardscholten73**
Hello @alexziskind1 ,

Another pull request for you to review.

In codeExchangeWithUrlCompletion:
The check on responseError made little sense.
The call to completion would always have null or undefined for the responseError
Also the else if construction would always result in the same call to completion (with both result and  responseError being null/undefined)

In completeLoginWithTokenResponseError:
There was no check if the authState is null/undefined
Sometimes resulting in an null reference exception.
I believe this is related to #50 

We are still getting this error, mainly when the user is already logged in on the provider (single sign on)
The browser immediately returns a token without user interaction. This sometimes results in the authState not being filled (yet?). I mentioned this before and believe it is some sort of race condition where the reply comes before the request is fully handled (or something like that)
Having difficulty following the flow with android an fully understanding what is happening in what sequence...
I added some comment to the code compare, hope it clarifies things.

Anyway, this pull request should solve some of the issues and improve things.

Hope you find the time to review this and make a new release.
If you have questions or want to discuss the issues please contact me.

kind regards,
Richard Scholten

## PR Checklist

- [ ] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [ ] All existing tests are passing
- [ ] Tests for the changes are included

## What is the new behavior?

Fixes/Implements/Closes #[Issue Number].

### CODE REVIEW

File: src/tns-oauth-login-sub-controller.ts [hunk 1]
```diff
@@ -131,29 +127,15 @@ export class TnsOAuthLoginSubController {
         response: HttpResponse,
         responseError: Error
       ) => {
-        if (!responseError) {
```
Comment (richardscholten73) [inline src/tns-oauth-login-sub-controller.ts]: first check if responseError is null/undefined

File: src/tns-oauth-login-sub-controller.ts [hunk 2]
```diff
@@ -131,29 +127,15 @@ export class TnsOAuthLoginSubController {
...
-              completion(tokenResult, responseError);
-            }
-          } else if (response.statusCode === 400) {
-            console.error("400 ERRROR Occurred");
-            // A 400 error can be due to an malformed request to the server by the SDK OR a malicious caller
-            // with an invalid auth code. At this point the server intentionally omit the detailed reason
-            // of the 400. We always assume the 400 is caused by a malicious caller and ignore it silently.
-            // Handling 400 error and notifying the user would mess up the auth flow and make the app
-            // less secure.
-            completion(null, responseError);
```
Comment (richardscholten73) [inline src/tns-oauth-login-sub-controller.ts]: responseError will always be null or undefined

File: src/tns-oauth-login-sub-controller.ts [hunk 3]
```diff
@@ -131,29 +127,15 @@ export class TnsOAuthLoginSubController {
...
-          } else if (response.statusCode > 400) {
-            // A non-400 error is unlikely to be caused by a malicious caller with an invalid auth code.
-            // So we don't ignore such error.
-
-            // Set responseError
-            completion(null, responseError);
```
Comment (richardscholten73) [inline src/tns-oauth-login-sub-controller.ts]: will alwas return same result as the else if above

**Comment by richardscholten73** [comment]
@alexziskind1, pending checks hangs.
It says build in progress, but actually finished yesterday...

**Comment by vincentpalita** [comment]
@richardscholten73 that's great!! Does this PR allows to not have anymore any loginCompletion of null error?
I opened another PR to allow to cancel the login operation on Android when the users closes the view or uses back button. 
I think both our PRs would stabilize the whole login process.

**Comment by richardscholten73** [comment]
> @richardscholten73 that's great!! Does this PR allows to not have anymore any loginCompletion of null error?
> I opened another PR to allow to cancel the login operation on Android when the users closes the view or uses back button.
> I think both our PRs would stabilize the whole login process.

Yes it does. However it does not solve the problem of loginCompletion being null sometimes. It just means it is handled better.

**Comment by richardscholten73** [comment]
@alexziskind1 Did you find the time to review the pull requests? Is there anything i can do to help?

**Comment by alexziskind1** [comment]
The PRs you've created have been merged @richardscholten73 . Thanks for the help!

**Comment by richardscholten73** [comment]
> The PRs you've created have been merged @richardscholten73 . Thanks for the help!

Thanks!

**Pull request merged**